src/main/java/gtPlusPlus/xmod/gregtech/api/metatileentity/implementations/base/GregtechMeta_MultiBlockBase.java

```diff
@@ -473,306 +477,203 @@ public int canBufferOutputs(ItemStack[] aItemOutputs, FluidStack[] aFluidOutputs
-                                // Remove fluid from output list, merge success
-                                aOutputFluids.remove(aOutputFluids.get(j));
-                                j--;
-                                // Remove hatch from hatch list, data is now invalid.
-                                aOutputHatches.remove(aOutputHatches.get(i));
-                                i--;
-                                // Re-add hatch to hatch list, with new data.
-                                Triplet<GT_MetaTileEntity_Hatch_Output, FluidStack, Integer> aNewHatchData =
-                                        new Triplet<GT_MetaTileEntity_Hatch_Output, FluidStack, Integer>(
-                                                aHatch, aNewHatchStack, aNewHatchStack.amount);
-                                aOutputHatches.add(aNewHatchData);
-                                // Check next fluid
-                                continue;
-                            }
-
-                        } else if (aHatchLockedFluid == null
-                                || aHatchLockedFluid.equals(
-                                        aOutputFluids.get(j).getFluid().getName())) {
-                            aEmptyFluidHatches++;
-                        } else {
-                            continue;
-                        }
+                // this fluid is not prevented by restrictions on output hatch
+                if (tHatch.getFluidAmount() == 0 || GT_Utility.areFluidsEqual(tHatch.getFluid(), tFluidOutput)) {
+                    // empty or same fluid - in any case, can accept
+                    if (tSpaceLeft >= tFluidOutput.amount) {
+                        // enough to hold all fluid
+                        tSpaceLeft -= tFluidOutput.amount;
+                        iterator.remove();
+                    } else {
+                        tFluidOutput.amount -= tSpaceLeft;
+                        break;
                     }
                 }
             }
+            // at this point we have either visited all output or is completed empty, so we can go over to next
+            // before that, check if we have handled all fluid, if yes, bail out for early exit.
+            if (tFluidsLeft.isEmpty()) break;
+        }
 
-            for (Triplet<GT_MetaTileEntity_Hatch_Output, FluidStack, Integer> aFreeHatchCheck : aOutputHatches) {
-                // Free Hatch
-                if ((aFreeHatchCheck.getValue_2() == null
-                                || aFreeHatchCheck.getValue_3() == 0
-                                || aFreeHatchCheck.getValue_1().getFluid() == null)
-                        && !aFreeHatchCheck.getValue_1().isFluidLocked()) {
-                    aEmptyFluidHatches++;
-                }
-            }
+        // check non-restrictive hatches
+        for (GT_MetaTileEntity_Hatch_Output tHatch : mOutputHatches) {
+            int tSpaceLeft = tHatch.getCapacity() - tHatch.getFluidAmount();
```

Comment:
Making second loop can be avoided by sorting mOutpuHatches. Its not really that big of a problem I guess

Reply:
I thought about sorting. But the fact I have to sort these every time this is called dissuade me from doing so, and I don't want to deviate from the original addOutputs() method in base class too much, as otherwise we run the risk of giving false negatives in some unforseen ways.

Reply:
Huh now when I think about it, the loop can be removed anyway. Order doesn't matter, the function only checks if you can output them. The output really happens in GT5 code and its outputting to the right hatches.

Reply:
Ideally I prefer to not have this function to return true if addOutputs would void.

--

The case in my head is when player are mixing hatches with different tier. Order can be significant in that case. I'm honestly not sure this case is valid, but I prefer to stay on the safe side until I'm willing to actually grab a pencil and paper to prove it.